types/cross-domain-utils/utils.d.ts

```diff
@@ -0,0 +1,247 @@
+
+/**
+ *  Returns true if the specified window has been opened with `window.open` (i.e. if it is a popup window)
+ */
+export function isPopup(win?: CrossDomainWindowType): boolean;
+
+/**
+ *  Returns true if the specified window has been opened as an iframe.
+ */
+export function isIframe(win?: CrossDomainWindowType): boolean;
+
+export function isFullpage(win?: CrossDomainWindowType): boolean;
+
+/**
+ *  Gets the numerical distance from the specified window to the top level window in that window's hierarchy.
+ *
+ *  - If the specified window is at the top, this will return 0.
+ */
+export function getDistanceFromTop(win: CrossDomainWindowType): number;
+
+/**
+ *  Gets the window `n` levels up from the specified window, if it exists.
+ */
+export function getNthParent(win: CrossDomainWindowType, n?: number): CrossDomainWindowType | null;
+
+export function getNthParentFromTop(win: CrossDomainWindowType, n?: number): CrossDomainWindowType;
+
+/**
+ *  Returns true if the windows are in the same hierarchy, with the same top level window
+ *
+ *  - Will return false if one of the windows is a popup and the other window is not a frame inside that popup.
+ */
+export function isSameTopWindow(win1: CrossDomainWindowType, win2: CrossDomainWindowType): boolean;
+
+/**
+ *  Returns true if the specified domain matches the pattern. The pattern can be one of:
+ *
+ *  - A literal string
+ *  - A regular expression
+ *  - An array of possible domains as strings
+ */
+export function matchDomain(pattern: DomainMatcher, origin: DomainMatcher): boolean;
+
+export function stringifyDomainPattern(pattern: DomainMatcher): string;
+
+/**
+ *  Get the full domain from the specified url, as a string.
+ *
+ *  - it will try to extract the domain from the url string if it starts with well known protocols (`http://`, `https://`, `file://`, and additionally `mock://` urls)
+ *  - if url string does not contain a known protocol, it will try to extract the domain calling `getDomain` using the current window as input
+ */
+export function getDomainFromUrl(url: string): string;
+
+/**
+ *  Calls the callback when the specified window closes, with checks running on the specified interval.
+ *
+ *  - Returns a listener object with a `.cancel()` method, to stop the loop
+ */
+export function onCloseWindow(win: CrossDomainWindowType, callback: (...args: any[]) => any, delay?: number, maxtime?: number): { cancel: () => void };
+
+/**
+ *  Returns true if the specified object is a window instance
+ */
+export function isWindow(obj: any): boolean;
+
+export function isBrowser(): boolean;
+
+export function isCurrentDomain(domain: string): boolean;
+
+export function isMockDomain(domain: string): boolean;
+
+export function normalizeMockUrl(url: string): string;
+
+export function closeWindow(win: CrossDomainWindowType): void;
+
+export function getFrameForWindow(win: CrossDomainWindowType): HTMLElement;
```

Comment:
this can return `undefined`:
https://github.com/krakenjs/cross-domain-utils/blob/87164b88ae8d2b59cf97608af9ce68efe66c2c16/src/utils.js#L1056
(hence `?` in flow I guess)